File: app/src/main/java/com/example/criminalintent/CrimeListFragment.java
package com.example.criminalintent;

import android.content.Context;
import android.os.Bundle;
import android.view.Menu;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.content.ContextCompat;
import androidx.core.os.LocaleListCompat;
import androidx.core.view.MenuHost;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.card.MaterialCardView;
import java.text.DateFormat;
import java.util.List;

public class CrimeListFragment extends Fragment {

    private static final String SAVED_SUBTITLE_VISIBLE = "subtitle";
    private RecyclerView mCrimeRecyclerView;
    private CrimeAdapter mAdapter;
    private boolean mSubtitleVisible;
    private Callbacks mCallbacks;

    private static final int VIEW_TYPE_NORMAL = 0;
    private static final int VIEW_TYPE_POLICE = 1;

    public interface Callbacks {
        void onCrimeSelected(Crime crime);
        void onCrimeDeleted(Crime crime);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mCallbacks = (Callbacks) context;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallbacks = null;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        MenuHost menuHost = requireActivity();
        menuHost.addMenuProvider(
            new MenuProvider() {
                @Override
                public void onCreateMenu(Menu menu, MenuInflater menuInflater) {
                    menuInflater.inflate(R.menu.fragment_crime_list, menu);
                }

                @Override
                public void onPrepareMenu(Menu menu) {
                    CrimeLab crimeLab = CrimeLab.get(requireActivity());
                    MenuItem newCrimeItem = menu.findItem(R.id.new_crime);
                    if (newCrimeItem != null) {
                        newCrimeItem.setVisible(crimeLab.canAddMoreCrimes());
                    }

                    MenuItem subtitleItem = menu.findItem(R.id.show_subtitle);
                    if (subtitleItem != null) {
                        subtitleItem.setTitle(mSubtitleVisible
                                ? R.string.hide_subtitle
                                : R.string.show_subtitle);
                    }
                }

                @Override
                public boolean onMenuItemSelected(MenuItem menuItem) {
                    if (menuItem.getItemId() == R.id.new_crime) {
                        CrimeLab crimeLab = CrimeLab.get(requireActivity());
                        if (!crimeLab.canAddMoreCrimes()) {
                            requireActivity().invalidateOptionsMenu();
                            return true;
                        }

                        Crime crime = new Crime();
                        crimeLab.addCrime(crime);
                        updateUI();
                        mCallbacks.onCrimeSelected(crime);
                        return true;
                    } else if (menuItem.getItemId() == R.id.show_subtitle) {
                        mSubtitleVisible = !mSubtitleVisible;
                        updateSubtitle();
                        requireActivity().invalidateOptionsMenu();
                        return true;
                    } else if (menuItem.getItemId() == R.id.change_language) {
                        showLanguageSelectionDialog();
                        return true;
                    }
                    return false;
                }
            },
            getViewLifecycleOwner(),
            Lifecycle.State.RESUMED
        );
    }

    private void showLanguageSelectionDialog() {
        String[] languages = {"English", "Español"};
        String[] langTags = {"en", "es"};

        new AlertDialog.Builder(requireContext())
                .setTitle("Select Language")
                .setItems(languages, (dialog, which) -> {
                    LocaleListCompat appLocales = LocaleListCompat.forLanguageTags(langTags[which]);
                    AppCompatDelegate.setApplicationLocales(appLocales);
                })
                .show();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_crime_list, container, false);

        if (savedInstanceState != null) {
            mSubtitleVisible = savedInstanceState.getBoolean(SAVED_SUBTITLE_VISIBLE);
        }

        mCrimeRecyclerView = view.findViewById(R.id.crime_recycler_view);
        mCrimeRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        attachSwipeToDismiss();

        updateUI();

        return view;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(SAVED_SUBTITLE_VISIBLE, mSubtitleVisible);
    }

    @Override
    public void onResume() {
        super.onResume();
        updateUI();
    }

    public void updateUI() {
        CrimeLab crimeLab = CrimeLab.get(getActivity());
        List<Crime> crimes = crimeLab.getCrimes();

        if (mAdapter == null) {
            mAdapter = new CrimeAdapter(crimes);
            mCrimeRecyclerView.setAdapter(mAdapter);
        } else {
            mAdapter.setCrimes(crimes);
            mAdapter.notifyDataSetChanged();
        }
        updateSubtitle();
        requireActivity().invalidateOptionsMenu();
    }

    private void attachSwipeToDismiss() {
        ItemTouchHelper.SimpleCallback swipeCallback =
                new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {
                    @Override
                    public boolean onMove(
                            RecyclerView recyclerView,
                            RecyclerView.ViewHolder viewHolder,
                            RecyclerView.ViewHolder target
                    ) {
                        return false;
                    }

                    @Override
                    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                        int position = viewHolder.getBindingAdapterPosition();
                        if (position == RecyclerView.NO_POSITION || mAdapter == null) {
                            if (mAdapter != null) {
                                mAdapter.notifyDataSetChanged();
                            }
                            return;
                        }

                        Crime crime = mAdapter.getCrime(position);
                        CrimeLab.get(requireContext()).deleteCrime(crime);
                        mCallbacks.onCrimeDeleted(crime);
                        updateUI();
                    }
                };

        new ItemTouchHelper(swipeCallback).attachToRecyclerView(mCrimeRecyclerView);
    }

    private void updateSubtitle() {
        int crimeCount = CrimeLab.get(getActivity()).getCrimeCount();
        String subtitle = getString(R.string.subtitle_format, crimeCount);
        if (!mSubtitleVisible) {
            subtitle = null;
        }

        AppCompatActivity activity = (AppCompatActivity) getActivity();
        if (activity.getSupportActionBar() != null) {
            activity.getSupportActionBar().setSubtitle(subtitle);
        }
    }

    private String getDisplayTitle(Crime crime) {
        String title = crime.getTitle();
        return title == null || title.trim().isEmpty()
                ? getString(R.string.crime_default_title)
                : title;
    }

    private CharSequence formatCrimeDate(Crime crime) {
        DateFormat dateFormat = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.SHORT);
        return dateFormat.format(crime.getDate());
    }

    private void bindStatusBadge(TextView statusTextView, Crime crime) {
        if (crime.isSolved()) {
            statusTextView.setText(R.string.crime_status_solved);
            statusTextView.setBackgroundResource(R.drawable.bg_badge_solved);
        } else {
            statusTextView.setText(R.string.crime_status_open);
            statusTextView.setBackgroundResource(R.drawable.bg_badge_open);
        }
    }

    private void bindCaseAppearance(
            MaterialCardView cardView,
            View caseStrip,
            Crime crime,
            int defaultStrokeColorRes,
            int defaultStripColorRes
    ) {
        int strokeColorRes = crime.isSolved() ? R.color.crime_accent_green : defaultStrokeColorRes;
        int stripColorRes = crime.isSolved() ? R.color.crime_accent_green : defaultStripColorRes;

        cardView.setStrokeColor(ContextCompat.getColor(cardView.getContext(), strokeColorRes));
        caseStrip.setBackgroundColor(ContextCompat.getColor(caseStrip.getContext(), stripColorRes));
    }

    private void bindSolvedIndicator(ImageView solvedImageView, Crime crime) {
        if (crime.isSolved()) {
            solvedImageView.setVisibility(View.VISIBLE);
            solvedImageView.setImageResource(R.drawable.ic_solved);
        } else {
            solvedImageView.setVisibility(View.GONE);
        }
    }

    private class CrimeHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private Crime mCrime;
        private final MaterialCardView cardView;
        private final View caseStrip;
        private final TextView titleTextView;
        private final TextView dateTextView;
        private final TextView statusTextView;
        private final ImageView solvedImageView;

        public CrimeHolder(View view) {
            super(view);
            cardView = (MaterialCardView) itemView;
            caseStrip = itemView.findViewById(R.id.case_strip);
            titleTextView = itemView.findViewById(R.id.crime_title);
            dateTextView = itemView.findViewById(R.id.crime_date);
            statusTextView = itemView.findViewById(R.id.crime_status_badge);
            solvedImageView = itemView.findViewById(R.id.crime_solved);
            itemView.setOnClickListener(this);
        }

        public void bind(Crime crime) {
            mCrime = crime;
            titleTextView.setText(getDisplayTitle(crime));
            dateTextView.setText(formatCrimeDate(crime));
            bindCaseAppearance(cardView, caseStrip, crime, R.color.crime_stroke, R.color.crime_accent_amber);
            bindStatusBadge(statusTextView, crime);
            bindSolvedIndicator(solvedImageView, crime);
        }

        @Override
        public void onClick(View v) {
            mCallbacks.onCrimeSelected(mCrime);
        }
    }

    private class CrimePoliceHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private Crime mCrime;
        private final MaterialCardView cardView;
        private final View caseStrip;
        private final TextView titleTextView;
        private final TextView dateTextView;
        private final TextView statusTextView;
        private final Button contactPoliceButton;
        private final ImageView solvedImageView;

        public CrimePoliceHolder(View view) {
            super(view);
            cardView = (MaterialCardView) itemView;
            caseStrip = itemView.findViewById(R.id.case_strip);
            titleTextView = itemView.findViewById(R.id.crime_title);
            dateTextView = itemView.findViewById(R.id.crime_date);
            statusTextView = itemView.findViewById(R.id.crime_status_badge);
            contactPoliceButton = itemView.findViewById(R.id.contact_police_button);
            solvedImageView = itemView.findViewById(R.id.crime_solved);
            itemView.setOnClickListener(this);
        }

        public void bind(Crime crime) {
            mCrime = crime;
            String displayTitle = getDisplayTitle(crime);
            titleTextView.setText(displayTitle);
            dateTextView.setText(formatCrimeDate(crime));
            bindCaseAppearance(cardView, caseStrip, crime, R.color.crime_accent_red, R.color.crime_accent_red);
            bindStatusBadge(statusTextView, crime);
            bindSolvedIndicator(solvedImageView, crime);

            contactPoliceButton.setOnClickListener(v -> 
                Toast.makeText(
                    getContext(),
                    getString(R.string.contact_police_message, displayTitle),
                    Toast.LENGTH_SHORT
                ).show()
            );
        }

        @Override
        public void onClick(View v) {
            mCallbacks.onCrimeSelected(mCrime);
        }
    }

    private class CrimeAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
        private List<Crime> crimes;

        public CrimeAdapter(List<Crime> crimes) {
            this.crimes = crimes;
        }

        public void setCrimes(List<Crime> crimes) {
            this.crimes = crimes;
        }

        public Crime getCrime(int position) {
            return crimes.get(position);
        }

        @Override
        public int getItemCount() {
            return crimes.size();
        }

        @Override
        public int getItemViewType(int position) {
            return crimes.get(position).isRequiresPolice() ? VIEW_TYPE_POLICE : VIEW_TYPE_NORMAL;
        }

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
            if (viewType == VIEW_TYPE_POLICE) {
                View view = layoutInflater.inflate(R.layout.list_item_crime_police, parent, false);
                return new CrimePoliceHolder(view);
            } else {
                View view = layoutInflater.inflate(R.layout.list_item_crime, parent, false);
                return new CrimeHolder(view);
            }
        }

        @Override
        public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
            Crime crime = crimes.get(position);
            if (holder instanceof CrimePoliceHolder) {
                ((CrimePoliceHolder) holder).bind(crime);
            } else if (holder instanceof CrimeHolder) {
                ((CrimeHolder) holder).bind(crime);
            }
        }
    }
}

File: app/src/main/java/com/bignerdranch/android/criminalintent/CrimeListFragment.kt
package com.bignerdranch.android.criminalintent

import android.app.AlertDialog
import android.content.Context
import android.os.Bundle
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat
import androidx.core.os.LocaleListCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bignerdranch.android.criminalintent.databinding.FragmentCrimeListBinding
import com.bignerdranch.android.criminalintent.databinding.ListItemCrimeBinding
import com.bignerdranch.android.criminalintent.databinding.ListItemCrimePoliceBinding

private const val VIEW_TYPE_NORMAL = 0
private const val VIEW_TYPE_POLICE = 1
private const val SAVED_SUBTITLE_VISIBLE = "subtitle"
private const val MAX_CRIMES = 10
private const val DATE_FORMAT = "EEE, MMM dd, yyyy"

class CrimeListFragment : Fragment() {

    /**
     * Required interface for hosting activities.
     */
    interface Callbacks {
        fun onCrimeSelected(crime: Crime)
    }

    private var callbacks: Callbacks? = null
    private var _binding: FragmentCrimeListBinding? = null
    private val binding get() = _binding!!

    private var adapter: CrimeAdapter? = null
    private var isSubtitleVisible: Boolean = false

    private val crimeListViewModel: CrimeListViewModel by lazy {
        ViewModelProvider(this)[CrimeListViewModel::class.java]
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callbacks = context as Callbacks?
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentCrimeListBinding.inflate(inflater, container, false)

        binding.crimeRecyclerView.layoutManager = LinearLayoutManager(context)
        
        binding.addCrimeButton.setOnClickListener {
            createNewCrime()
        }

        if (savedInstanceState != null) {
            isSubtitleVisible = savedInstanceState.getBoolean(SAVED_SUBTITLE_VISIBLE)
        }

        setupItemTouchHelper()
        
        return binding.root
    }

    private fun setupItemTouchHelper() {
        val itemTouchHelperCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean = false

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                adapter?.crimes?.get(position)?.let { crime ->
                    CrimeLab.get(requireContext()).deleteCrime(crime)
                    crimeListViewModel.loadCrimes(requireContext())
                }
            }
        }

        ItemTouchHelper(itemTouchHelperCallback).attachToRecyclerView(binding.crimeRecyclerView)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        crimeListViewModel.crimes.observe(viewLifecycleOwner) { crimes ->
            crimes?.let {
                updateUI(it)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        crimeListViewModel.loadCrimes(requireContext())
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(SAVED_SUBTITLE_VISIBLE, isSubtitleVisible)
    }

    override fun onDetach() {
        super.onDetach()
        callbacks = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.fragment_crime_list, menu)

        val subtitleItem = menu.findItem(R.id.show_subtitle)
        if (isSubtitleVisible) {
            subtitleItem.setTitle(R.string.hide_subtitle)
        } else {
            subtitleItem.setTitle(R.string.show_subtitle)
        }

        val newCrimeItem = menu.findItem(R.id.new_crime)
        val crimeCount = crimeListViewModel.crimes.value?.size ?: 0
        newCrimeItem.isVisible = crimeCount < MAX_CRIMES
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.new_crime -> {
                createNewCrime()
                true
            }
            R.id.show_subtitle -> {
                isSubtitleVisible = !isSubtitleVisible
                activity?.invalidateOptionsMenu()
                updateSubtitle()
                true
            }
            R.id.change_language -> {
                showLanguageDialog()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showLanguageDialog() {
        val languages = arrayOf(getString(R.string.english), getString(R.string.spanish))
        val languageCodes = arrayOf("en", "es")

        AlertDialog.Builder(requireContext())
            .setTitle(R.string.change_language)
            .setItems(languages) { _, which ->
                val appLocale: LocaleListCompat = LocaleListCompat.forLanguageTags(languageCodes[which])
                AppCompatDelegate.setApplicationLocales(appLocale)
            }
            .show()
    }

    private fun createNewCrime() {
        val crimeCount = crimeListViewModel.crimes.value?.size ?: 0
        if (crimeCount >= MAX_CRIMES) {
            return
        }
        val crime = Crime()
        CrimeLab.get(requireContext()).addCrime(crime)
        crimeListViewModel.loadCrimes(requireContext())
        callbacks?.onCrimeSelected(crime)
    }

    private fun updateSubtitle() {
        val crimeCount = crimeListViewModel.crimes.value?.size ?: 0
        val subtitle = if (isSubtitleVisible) {
            resources.getQuantityString(R.plurals.subtitle_plural, crimeCount, crimeCount)
        } else {
            null
        }

        val activity = activity as AppCompatActivity
        activity.supportActionBar?.subtitle = subtitle
    }

    fun updateUI() {
        crimeListViewModel.loadCrimes(requireContext())
    }

    private fun updateUI(crimes: List<Crime>) {
        val isEmpty = crimes.isEmpty()
        binding.crimeRecyclerView.visibility = if (isEmpty) View.GONE else View.VISIBLE
        binding.emptyView.visibility = if (isEmpty) View.VISIBLE else View.GONE

        if (adapter == null) {
            adapter = CrimeAdapter(crimes)
            binding.crimeRecyclerView.adapter = adapter
        } else {
            adapter?.crimes = crimes
            adapter?.notifyDataSetChanged()
        }

        activity?.invalidateOptionsMenu()
        updateSubtitle()
    }

    private abstract inner class BaseCrimeHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        protected lateinit var crime: Crime
        
        init {
            itemView.setOnClickListener(this)
        }

        abstract fun bind(crime: Crime)

        override fun onClick(v: View) {
            callbacks?.onCrimeSelected(crime)
        }

        protected fun applySolvedStyling(titleView: android.widget.TextView, dateView: android.widget.TextView, isSolved: Boolean) {
            val color = if (isSolved) {
                ContextCompat.getColor(itemView.context, R.color.green)
            } else {
                ContextCompat.getColor(itemView.context, R.color.black)
            }
            titleView.setTextColor(color)
            dateView.setTextColor(color)
        }
    }

    private inner class CrimeHolder(private val itemBinding: ListItemCrimeBinding) : BaseCrimeHolder(itemBinding.root) {
        override fun bind(crime: Crime) {
            this.crime = crime
            itemBinding.crimeTitle.text = crime.title
            itemBinding.crimeDate.text = DateFormat.format(DATE_FORMAT, crime.date)
            itemBinding.crimeSolved.visibility = if (crime.isSolved) View.VISIBLE else View.GONE
            applySolvedStyling(itemBinding.crimeTitle, itemBinding.crimeDate, crime.isSolved)
        }
    }

    private inner class PoliceCrimeHolder(private val itemBinding: ListItemCrimePoliceBinding) : BaseCrimeHolder(itemBinding.root) {
        override fun bind(crime: Crime) {
            this.crime = crime
            itemBinding.crimeTitle.text = crime.title
            itemBinding.crimeDate.text = DateFormat.format(DATE_FORMAT, crime.date)
            itemBinding.crimeSolved.visibility = if (crime.isSolved) View.VISIBLE else View.GONE
            itemBinding.contactPoliceButton.setOnClickListener {
                Toast.makeText(context, "Police contacted for ${crime.title}", Toast.LENGTH_SHORT).show()
            }
            applySolvedStyling(itemBinding.crimeTitle, itemBinding.crimeDate, crime.isSolved)
        }
    }

    private inner class CrimeAdapter(var crimes: List<Crime>) : RecyclerView.Adapter<BaseCrimeHolder>() {

        override fun getItemViewType(position: Int): Int {
            return if (crimes[position].requiresPolice) VIEW_TYPE_POLICE else VIEW_TYPE_NORMAL
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseCrimeHolder {
            val layoutInflater = LayoutInflater.from(activity)
            return if (viewType == VIEW_TYPE_POLICE) {
                PoliceCrimeHolder(ListItemCrimePoliceBinding.inflate(layoutInflater, parent, false))
            } else {
                CrimeHolder(ListItemCrimeBinding.inflate(layoutInflater, parent, false))
            }
        }

        override fun onBindViewHolder(holder: BaseCrimeHolder, position: Int) {
            holder.bind(crimes[position])
        }

        override fun getItemCount() = crimes.size
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        fun newInstance(): CrimeListFragment {
            return CrimeListFragment()
        }
    }
}
